File: files/400-499/440/CH-440.R
library(tidyverse)
library(readxl)
library(igraph)

path <- "400-499/440/CH-440 - Knight Distance Problem.xlsx"
start <- "A2"
end <- "F7"
blocked <- expand_grid(
  file = c("C", "D"),
  rank = 1:6
) %>%
  transmute(cell = paste0(file, rank)) %>%
  pull(cell)
test <- read_excel(path, range = "L3:L6") %>%
  mutate(`example Paths` = str_replace_all(`example Paths`, " ", ""))

blocked <- c(outer(c("C", "D"), 1:6, paste0))
cells <- expand_grid(x = 1:8, y = 1:8) %>%
  mutate(cell = paste0(LETTERS[x], y)) %>%
  filter(!cell %in% blocked)
moves <- expand_grid(
  dx = c(-2, -1, 1, 2),
  dy = c(-2, -1, 1, 2)
) %>%
  filter(abs(dx) + abs(dy) == 3)
edges <- tidyr::crossing(
  cells %>% rename(x1 = x, y1 = y, from = cell),
  moves
) %>%
  transmute(
    from,
    x2 = x1 + dx,
    y2 = y1 + dy
  ) %>%
  inner_join(
    cells %>% rename(x2 = x, y2 = y, to = cell),
    by = c("x2", "y2")
  ) %>%
  select(from, to)
g <- graph_from_data_frame(edges, directed = FALSE)
sp <- all_shortest_paths(
  g,
  from = "A2",
  to = "F7"
)$res
paths <- map(sp, as_ids) %>%
  unique() %>%
  purrr::map_chr(., paste, collapse = "->")

paths # there are 2 identical (by sight).

File: files/400-499/440/CH-440.py
import pandas as pd
import networkx as nx
from itertools import product

path = "400-499/440/CH-440 - Knight Distance Problem.xlsx"
test = pd.read_excel(path, usecols="L", skiprows=2, nrows=4).rename(
    columns=lambda x: x.strip()
)
test["example Paths"] = test["example Paths"].str.replace(" ", "", regex=False)
start = "A2"
end = "F7"
blocked = {f"{file}{rank}" for file, rank in product("CD", range(1, 7))}
cells = {f"{file}{rank}" for file, rank in product("ABCDEFGH", range(1, 9))} - blocked
moves = [
    (dx, dy) for dx, dy in product((-2, -1, 1, 2), repeat=2) if abs(dx) + abs(dy) == 3
]


def move(cell, dx, dy):
    x = ord(cell[0]) - ord("A") + dx
    y = int(cell[1]) + dy
    if 0 <= x < 8 and 1 <= y <= 8:
        return f"{chr(ord('A') + x)}{y}"


g = nx.Graph()
g.add_nodes_from(cells)
g.add_edges_from(
    (cell, target)
    for cell in cells
    for dx, dy in moves
    if (target := move(cell, dx, dy)) in cells
)
paths = ["->".join(path) for path in nx.all_shortest_paths(g, start, end)]
paths = list(dict.fromkeys(paths))
